CRM/Core/Payment.php

```diff
@@ -1186,9 +1186,22 @@ protected function doDirectPayment(&$params) {
    * Usage:
    * Payment processors should override this function directly instead of using doDirectPayment/doTransferCheckout which are deprecated.
    * Payment processors should set and return payment_status_id (Pending if the IPN will complete it, Completed if successful).
-   * @fixme For the contribution workflow we have a contributionID, but for the event and membership workflow the contribution has not yet been created
-   *  so we can't update params directly on the contribution.  However if you return trxn_id, fee_amount, net_amount they will be set on the contribution
-   *  by those workflows.  Ideally all workflows would create a pending contribution BEFORE calling doPayment (eg. https://github.com/civicrm/civicrm-core/pull/13763 for events)
+   * @fixme Creating a contribution record is inconsistent! We should always create a contribution BEFORE calling doPayment. However currently:
+   *  * Contribution Pages: Creates a contribution before calling doPayment.
+   *  * Contribution Pages with Membership: Does NOT create a contribution. UNLESS auto-renew=TRUE in which case both a recurring contribution
```

Comment:
@mattwire I think that when there is no recurring it does?

Reply:
@eileenmcnaughton When there is no recurring what does?  contribution page with membership does not create a contribution if auto renew is disabled.

Reply:
Hmm - I'd have to spend a while digging to check but we have 
- membership with separate contribution
- membership without separate contribution
- the above with combinations of renew

My other worry about documenting the gaps here is we might not remember to remove them - perhaps we should instead be more aggressively documenting the places that should, but don't, create pending contributions or using an issue in gitlab to track the last status of the meta project 'make all places create pending & then call doPayment & then call Payment.create' and just point to that from here?

Reply:
@eileenmcnaughton Every time I work on / develop a payment processor I have to go through and guess what might be returned and what should happen.  I'm proposing adding these code comments because it's the only place a developer will actually find them!  And better to start by documentation what is actually happening (with lot's of @fixme tags indicating this is not the preferred way).  If we were going to fix all the "known" and the "unknown" workflows in the next release I'd agree, but realistically it's a year or two away and in the mean time we still need payment processors to try and actually work.